src/Parsers/IAST.h

```diff
@@ -157,7 +157,23 @@ class IAST : public std::enable_shared_from_this<IAST>, public TypePromotion<IAS
             set(field, child);
     }
 
-    void reset(IAST *& field);
+    template <typename T>
+    void reset(T * & field)
```

Comment:
btw why need to add template? I see you now use it for ASTStorage, but it is also IAST

Reply:
Since you can't bind different type (even though it is derived from the base class) to **non const reference** for function argument - https://godbolt.org/z/hP6z6rKMs (since you may change it to other type for example and this will not be valid)

And to make it more like other IAST interface I've maked it templated by AST type.